api/templates/design.go

```diff
@@ -6,13 +6,17 @@ import (
 	. "goa.design/goa/v3/dsl"
```

Comment:
V súbore [api/templates/design.go](https://github.com/keboola/keboola-as-code/blob/05dfc63a57578d41650eb54f338feb8aae4649c1/api/templates/design.go) je definované Templates API:
- Pouziva sa kniznia Goa: https://goa.design/
- Je to trocha taky experiment, ktory sa celkom podaril.
- Vieme tak rychlo robit zmeny v API, naraz menime vsetky definicie a dostaneme vygenerovany kod.
- Pri `make generate-code` sa z tohto suboru vygeneruju vsetky struktury pre hodnoty (in/out payload ...) a kopa dalsieho kodu.
- Generovany kod je postaveny na standardnej kniznici v Go `net/http`, takze sa na to jednoducho napajaju dalsie middlewares, DataDog ....
- My uz potom iba musime implementovat vygenerovany `Service` interface.
- Implementacia je tu: https://github.com/keboola/keboola-as-code/blob/main/internal/pkg/api/server/templates/service/service.go
- Vsimni si, ze endpoint handler uz dostane priamo vsetky potrebne zavislosti `dependencies.ForPublicRequest/dependencies.ForProjectRequest` takze je to dobre testovatelne/pochopitelne (ak ked to testujeme len E2E, co nam dava vacsi zmysel).
- Ak nieco Goa kniznica nepodporuje, alebo si to predstavujeme inak, tak si vytvorime extension:
  - https://github.com/keboola/keboola-as-code/tree/05dfc63a57578d41650eb54f338feb8aae4649c1/internal/pkg/api/server/templates/extension
  - Napr. tie zavislosti v endpoint handler su toto extension: https://github.com/keboola/keboola-as-code/blob/05dfc63a57578d41650eb54f338feb8aae4649c1/internal/pkg/api/server/templates/extension/dependencies/dependencies.go
  - Generovanie kodu funguje na baze sablon (podobnych Jinja) - `text/template` pkg.
  - Extensions su kratke ale zlozite :slightly_smiling_face: , je potrebne si nastudovat sablonu co tam je a strukturu dat, ktora do nej ide.

Reply:
Dívali jste se i na nějaké code-first openapi tools? Dost mě tohle zajímá, ale nikde jsem zatím nenašel dobrý comparison code-first a design-first. Jen mám nepotvrzený domnění, že code-first má menší maintenance load (nemusíš vůbec řešit separatní "design" fázi, jen píšeš kód a z toho kódu máš vygenerovaný spec). Chtěl bych tohle pochopit, rád bych se o tom nějak víc pobavil i mimo PR commments 😄.

Reply:
(priklad smithy: https://github.com/wasmCloud/examples/blob/main/interface/payments/payments.smithy)

Reply:
este k (mozeme prebrat na standupe)
> nemusíš vůbec řešit separatní "design" fázi, jen píšeš kód 

pisanie kodu by som videl ako 2 casti:
1. pisanie kostry kodu (mam funkcne API, ktore nic nerobi)
2. pisanie implementacie/ business logiky

Pri `code-first` riesis bod 1. a 2.

Pri `design-first` riesis bod 1. vramci `design` a potom bod `2.`
 - Hned ako dopises `design.go/smithy file` a spustit generovanie -> tak mas funkcne API, ktore nic nerobi.